src/shogun/preprocessor/RandomFourierGaussPreproc.cpp

```diff
@@ -7,394 +7,88 @@
-		throw ShogunException(
-				"float64_t * CRandomFourierGaussPreproc::apply_to_feature_vector(...): test_rfinited()==false: you need to call before CRandomFourierGaussPreproc::init (Features *f) OR	1. set_dim_feature_space(const int32 dim), 2. set_dim_input_space(const int32 dim), 3. init_randomcoefficients() or set_randomcoefficients(...) \n");
-	}
-
-	float64_t val = std::sqrt(2.0 / cur_dim_feature_space);
-	SGVector<float64_t> res(cur_dim_feature_space);
-
-	for (int32_t od = 0; od < cur_dim_feature_space; ++od) {
-		SGVector<float64_t> wrapper(randomcoeff_multiplicative+od*cur_dim_input_space, cur_dim_input_space, false);
-		res[od] = val * cos(randomcoeff_additive[od] + linalg::dot(vector, wrapper));
-	}
-
-	return res;
+	return static_cast<SGVector<float64_t>>(
+	    apply_to_matrix(static_cast<SGMatrix<float64_t>>(vector)));
 }
 
-SGMatrix<float64_t>
-RandomFourierGaussPreproc::apply_to_matrix(SGMatrix<float64_t> matrix)
+SGMatrix<float64_t> RandomFourierGaussPreproc::apply_to_matrix(SGMatrix<float64_t> matrix)
 {
-	// version for case dim_feature_space < dim_input space with direct transformation on feature matrix ??
-
 	auto num_vectors = matrix.num_cols;
 	auto num_features = matrix.num_rows;
 
-	io::info("get Feature matrix: {}x{}", num_vectors, num_features);
+	assert_fitted();
+	SG_DEBUG("Got Feature matrix: {}x{}", num_vectors, num_features);
 
-	if (num_features != cur_dim_input_space)
-	{
-		throw ShogunException(
-		    "float64_t * "
-		    "CRandomFourierGaussPreproc::apply_to_matrix("
-		    "SGMatrix<float64_t> matrix): matrix.num_rows != "
-		    "cur_dim_input_space is not allowed\n");
-	}
+	SGMatrix<float64_t> projection(m_dim_output, num_vectors);
 
-	SGMatrix<float64_t> res(cur_dim_feature_space, num_vectors);
+	linalg::matrix_prod(m_basis,matrix,projection);
+	linalg::add_vector(projection, m_offset, projection);
+	linalg::cos(projection, projection);
+	linalg::scale(projection, projection, std::sqrt(2.0 / m_dim_output));
 
-	auto val = std::sqrt(2.0 / cur_dim_feature_space);
-
-	for (auto vec : range(num_vectors))
-	{
-		for (auto od : range(cur_dim_feature_space))
-		{
-			SGVector<float64_t> a(
-			    matrix.matrix + vec * num_features, cur_dim_input_space, false);
-			SGVector<float64_t> b(
-			    randomcoeff_multiplicative + od * cur_dim_input_space,
-			    cur_dim_input_space, false);
-			res(od, vec) =
-			    val * cos(randomcoeff_additive[od] + linalg::dot(a, b));
-		}
-	}
+	return projection;	
+}
 
-	return res;
+SGMatrix<float64_t> RandomFourierGaussPreproc::sample_spectral_density(int32_t dim_input_space) const
+{
+	NormalDistribution<float64_t> normal_dist;
+	auto sampled_kernel = SGMatrix<float64_t>(m_dim_output, dim_input_space);
+	const auto width = std::exp(m_log_width * 2.0) * 2.0;
+	std::generate(
+	    sampled_kernel.begin(), sampled_kernel.end(),
+	    [this, &normal_dist, &width]() {
+		    auto coeff = std::sqrt(2.0 / width) * normal_dist(m_prng);
+		    return coeff;
+	    });
```

Comment:
this looks much nicer in my opinion! And it should reduce cache misses. Minor thing, why don't you just return the right hand side of `coeff`?

Reply:
btw why not store the full multiplier outside of the loop, including `std::sqrt(2.0 / width)` ?

Reply:
yes, that could be done. In this situation I would expect a good compiler to just compute it once, because width is const. But to be sure can just compute this outside the lambda

Reply:
++